lib/bulk/common.js

```diff
@@ -441,13 +444,519 @@ const BulkWriteError = function(error, result) {
+  if (op.insertOne && op.insertOne.document == null) {
+    if (forceServerObjectId !== true && op.insertOne._id == null) op.insertOne._id = new ObjectID();
+    return bulk.addToOperationsList(bulk, INSERT, op.insertOne);
+  } else if (op.insertOne && op.insertOne.document) {
+    if (forceServerObjectId !== true && op.insertOne.document._id == null)
+      op.insertOne.document._id = new ObjectID();
+    return bulk.addToOperationsList(bulk, INSERT, op.insertOne.document);
+  }
+
+  if (op.insertMany) {
+    for (var i = 0; i < op.insertMany.length; i++) {
+      if (forceServerObjectId !== true && op.insertMany[i]._id == null)
+        op.insertMany[i]._id = new ObjectID();
+      bulk.addToOperationsList(bulk, INSERT, op.insertMany[i]);
+    }
+
+    return;
+  }
+
+  // No valid type of operation
+  throw toError(
+    'bulkWrite only supports insertOne, insertMany, updateOne, updateMany, removeOne, removeMany, deleteOne, deleteMany'
+  );
+};
+
+/**
+ * Execute next write command in a chain
+ *
+ * @method
+ * @param {class} bulk either OrderedBulkOperation or UnorderdBulkOperation
+ * @param {object} writeConcern
+ * @param {object} options
+ * @param {function} callback
+ */
+const Execute = function(_writeConcern, options, callback, bulk) {
+  if (typeof options === 'function') (callback = options), (options = {});
+  options = options || {};
+
+  if (bulk.s.executed) {
+    var executedError = toError('batch cannot be re-executed');
+    return typeof callback === 'function'
+      ? callback(executedError, null)
+      : bulk.s.promiseLibrary.reject(executedError);
+  }
+
+  if (typeof _writeConcern === 'function') {
+    callback = _writeConcern;
+  } else if (_writeConcern && typeof _writeConcern === 'object') {
+    bulk.s.writeConcern = _writeConcern;
+  }
+
+  // If we have current batch
+  if (bulk.isOrdered()) {
+    if (bulk.s.currentBatch) bulk.s.batches.push(bulk.s.currentBatch);
+  } else {
+    if (bulk.s.currentInsertBatch) bulk.s.batches.push(bulk.s.currentInsertBatch);
+    if (bulk.s.currentUpdateBatch) bulk.s.batches.push(bulk.s.currentUpdateBatch);
+    if (bulk.s.currentRemoveBatch) bulk.s.batches.push(bulk.s.currentRemoveBatch);
+  }
+  // If we have no operations in the bulk raise an error
+  if (bulk.s.batches.length === 0) {
+    var emptyBatchError = toError('Invalid Operation, no operations specified');
+    return typeof callback === 'function'
+      ? callback(emptyBatchError, null)
+      : bulk.s.promiseLibrary.reject(emptyBatchError);
+  }
+  return {
+    options: options,
+    callback: callback
+  };
+};
+
+/**
+ * Manages final options before executing command
+ *
+ * @param {boolean} bool
+ * @param {object} options
+ * @param {number} batch
+ * @param {function} resultHandler
+ * @param {function} callback
+ * @param {class} bulk either OrderedBulkOperation or UnorderdBulkOperation
+ */
+const FinalOptionsManagement = function(bool, options, batch, resultHandler, callback, bulk) {
```

Comment:
1. I'm not a fan of the name `bool` for a variable. It doesn't really tell us much.
2. In general, anything more than 4 arguments to a function means it might be a good time to example why there are so man. Can this be refactored to make the # of parameters lower? Either by grouping them together into an object, or by breaking this functionality across multiple functions?